src/tests/Stream/SourceFilterTest.cpp

```diff
@@ -0,0 +1,282 @@
+        SomeRegistrar);
+  }
+};
+
+TEST_F(SourceFilterTest, InitState) {
+  auto UnderTest = getTestFilter();
+  EXPECT_FALSE(UnderTest->hasFinished());
+}
+using trompeloeil::_;
+
+class yyyyFbReader : public FileWriter::FlatbufferReader {
+public:
+  bool verify(FileWriter::FlatbufferMessage const &) const override {
+    return true;
+  }
+
+  std::string
+  source_name(FileWriter::FlatbufferMessage const &) const override {
+    return "some_name";
+  }
+
+  uint64_t timestamp(FileWriter::FlatbufferMessage const &) const override {
+    return yyyyFbReader::Timestamp;
+  }
+  static void setTimestamp(uint64_t NewTime) {
+    yyyyFbReader::Timestamp = NewTime;
+  }
+
+private:
+  static uint64_t Timestamp;
+};
+
+uint64_t yyyyFbReader::Timestamp{1};
+
+FileWriter::FlatbufferMessage generateMsg() {
+  std::array<uint8_t, 9> SomeData{'y', 'y', 'y', 'y', 'y', 'y', 'y', 'y', 'y'};
+  setExtractorModule<yyyyFbReader>("yyyy");
+  return FileWriter::FlatbufferMessage(SomeData.data(), SomeData.size());
+}
+
+TEST_F(SourceFilterTest, InvalidMessage) {
+  FORBID_CALL(Writer, addMessage(_));
+  auto UnderTest = getTestFilter();
+  FileWriter::FlatbufferMessage TestMsg;
+  UnderTest->filterMessage(std::move(TestMsg));
+  EXPECT_TRUE(UnderTest->MessagesDiscarded == 1);
+  EXPECT_FALSE(UnderTest->hasFinished());
+}
+
+TEST_F(SourceFilterTest, MessageNoDest) {
+  FORBID_CALL(Writer, addMessage(_));
+  auto UnderTest = getTestFilter();
+  auto TestMsg = generateMsg();
+  UnderTest->filterMessage(std::move(TestMsg));
+  EXPECT_FALSE(UnderTest->hasFinished());
+}
+
+TEST_F(SourceFilterTest, BufferedMessageBeforeStart) {
+  REQUIRE_CALL(Writer, addMessage(_)).TIMES(1);
+  auto UnderTest = getTestFilter();
+  UnderTest->addDestinationPtr(0);
+  auto TestMsg = generateMsg();
+  UnderTest->filterMessage(std::move(TestMsg));
+  EXPECT_TRUE(UnderTest->MessagesReceived == 1);
+  EXPECT_TRUE(UnderTest->MessagesTransmitted == 0);
+  EXPECT_FALSE(UnderTest->hasFinished());
+}
+
+TEST_F(SourceFilterTest, MultipleMessagesBeforeStart) {
+  REQUIRE_CALL(Writer, addMessage(_)).TIMES(1);
+  auto UnderTest = getTestFilter();
+  UnderTest->addDestinationPtr(0);
+  auto TestMsg = generateMsg();
+  UnderTest->filterMessage(std::move(TestMsg));
+  yyyyFbReader::setTimestamp(2);
+  auto TestMsg2 = generateMsg();
+  UnderTest->filterMessage(std::move(TestMsg2));
+  EXPECT_TRUE(UnderTest->MessagesReceived == 2);
+  EXPECT_TRUE(UnderTest->MessagesTransmitted == 0);
+  EXPECT_FALSE(UnderTest->hasFinished());
+}
+
+TEST_F(SourceFilterTest, SameTSBeforeStart) {
+  REQUIRE_CALL(Writer, addMessage(_)).TIMES(1);
+  auto UnderTest = getTestFilter();
+  UnderTest->addDestinationPtr(0);
+  auto TestMsg = generateMsg();
+  UnderTest->filterMessage(std::move(TestMsg));
+  TestMsg = generateMsg();
+  UnderTest->filterMessage(std::move(TestMsg));
+  EXPECT_TRUE(UnderTest->MessagesReceived == 2);
+  EXPECT_TRUE(UnderTest->RepeatedTimestamp == 1);
```

Comment:
What is the purpose of tracking repeated timestamps?

Reply:
See DM-1869. I think there is another JIRA ticket for this as well.